File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/DelayedActions.java
package org.firstinspires.ftc.teamcode.utils;

import com.sun.tools.javac.util.Pair;

import java.util.ArrayList;
import java.util.List;

/**
 * A class for running actions after a delay.
 */
public class DelayedActions {
    private final List<Pair<Long, Runnable>> actions;
    public DelayedActions() {
        actions = new ArrayList<>();
    }
    /**
     * Adds an action to be run after a delay.
     * @param delay the delay in seconds.
     * @param action the action to run.
     */
    public void addDelayed(double delay, Runnable action) {
        actions.add(new Pair<>(System.currentTimeMillis() + (long) (delay * 1000), action));
    }
    /**
     * Runs all actions that have passed their delay.
     */
    public void run() {
        long currentTime = System.currentTimeMillis();
        actions.removeIf(pair -> {
            if (pair.fst < currentTime) {
                pair.snd.run();
                return true;
            }
            return false;
        });
    }
    /**
     * Clears all actions.
     * @see #addDelayed(double, Runnable)
     * @noinspection UnusedDeclaration
     */
    public void clear() {
        actions.clear();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/utils/DelayedActions.kt
package org.firstinspires.ftc.utils

/**
 * A class for running actions after a delay.
 */
class DelayedActions {
    private val actions: MutableList<Pair<Double,() -> Unit>> = mutableListOf()
    /**
     * Adds an action to be run after a delay.
     * @param delay the delay in seconds.
     * @param action the action to run.
     */
    fun addDelay(delay: Double, action: () -> Unit) {
        actions.add(Pair(delay * 1000 + System.currentTimeMillis(), action))
    }
    /**
     * Runs all actions that have passed their delay.
     */
    fun run() {
        actions.forEach {
            if (it.first <= System.currentTimeMillis()) {
                it.second
                actions.remove(it)
            }
        }
    }
    /**
     * Clears all actions.
     * @see addDelay
     * @noinspection UnusedDeclaration
     */
    fun clear() {
        actions.clear()
    }
}